File: 2024/day5/solution.py
from collections import defaultdict, deque
import sys

sys.setrecursionlimit(10**6)
infile = sys.argv[1] if len(sys.argv) >= 2 else "test_inputs.txt"

p1 = 0
p2 = 0
D = open(infile).read().strip()

ER = defaultdict(set)
E = defaultdict(set)
edges, queries = D.split("\n\n")
for line in edges.split("\n"):
    x, y = line.split("|")
    x, y = int(x), int(y)
    E[y].add(x)
    ER[x].add(y)

for query in queries.split("\n"):
    vs = [int(x) for x in query.split(",")]
    ok = True
    for i, x in enumerate(vs):
        for j, y in enumerate(vs):
            if i < j and y in E[x]:
                ok = False
    if ok:
        p1 += vs[len(vs) // 2]
    else:
        good = []
        Q = deque([])
        D = {v: len(ER[v]) for v in vs}
        for v in vs:
            if D[v] == 0:
                Q.append(v)
        while Q:
            x = Q.popleft()
            good.append(x)
            for y in ER[x]:
                if y in D:
                    D[y] = -1
                    if D[y] == 0:
                        Q.append(y)
        p2 += good[len(good) // 2]


print(p2)

File: 2024/day5/solution.rs
use std::collections::{HashMap, HashSet};
use std::io::{self};

fn main() -> io::Result<()> {
    let infile = "inputs.txt";

    let contents = std::fs::read_to_string(infile)?;

    let parts: Vec<&str> = contents.split("\n\n").collect();
    let edges = parts[0];
    let queries = parts[1];

    let mut graph: HashMap<i32, HashSet<i32>> = HashMap::new();

    for line in edges.lines() {
        let parts: Vec<&str> = line.split('|').collect();
        let x: i32 = parts[0].parse().unwrap();
        let y: i32 = parts[1].parse().unwrap();

        graph.entry(y).or_insert_with(HashSet::new).insert(x);
    }
    let mut p1 = 0;
    for query in queries.lines() {
        let vs: Vec<i32> = query.split(',').map(|x| x.parse().unwrap()).collect();
        let mut ok = true;

        for i in 0..vs.len() {
            for j in (i + 1)..vs.len() {
                if graph
                    .get(&vs[i])
                    .unwrap_or(&HashSet::new())
                    .contains(&vs[j])
                {
                    ok = false;
                    break;
                }
            }
            if !ok {
                break;
            }
        }

        if ok {
            p1 += vs[vs.len() / 2];
        }
    }

    dbg!(&p1);
    Ok(())
}
